state/protocol/protocol_state/state/mutable_protocol_state_test.go

```diff
@@ -0,0 +1,650 @@
+	workingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Twice()
+	workingStateMachine.On("Build").Return(transaction.NewDeferredBlockPersist(), nil).Maybe()
+
+	exception := errors.New("exception")
+	failingStateMachine := *protocol_statemock.NewOrthogonalStoreStateMachine[protocol.KVStoreReader](s.T())
+	failingStateMachine.On("EvolveState", mock.MatchedBy(emptySlice[flow.ServiceEvent]())).Return(nil).Twice()
+	failingStateMachine.On("Build").Return(nil, exception).Once()
+
+	s.Run("failing state machine is last", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = workingStateMachine, failingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.False(s.T(), protocol.IsInvalidServiceEventError(err))
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+
+	failingStateMachine.On("Build").Return(nil, exception).Once()
+	s.Run("failing state machine is first", func() {
+		s.kvStateMachines[0], s.kvStateMachines[1] = failingStateMachine, workingStateMachine //nolint:govet
+		_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+		require.Error(s.T(), err)
+		require.False(s.T(), protocol.IsInvalidServiceEventError(err))
+		require.True(s.T(), dbUpdates.IsEmpty())
+	})
+}
+
+// Test_EncodeFailed tests that `stateMutator` returns an exception when encoding the resulting state fails.
+func (s *StateMutatorSuite) Test_EncodeFailed() {
+	exception := errors.New("exception")
+	s.protocolKVStoreDB = *protocol_statemock.NewProtocolKVStore(s.T())
+	s.protocolKVStoreDB.On("ByBlockID", s.candidate.ParentID).Return(&s.parentState, nil)
+
+	expectedResultingStateID := unittest.IdentifierFixture()
+	modifyState := func(_ mock.Arguments) {
+		s.evolvingState.On("ID").Return(expectedResultingStateID, nil).Once()
+	}
+	s.kvStateMachines[0] = s.mockStateTransition().ServiceEventsMatch(emptySlice[flow.ServiceEvent]()).DuringEvolveState(modifyState).Mock()
+	s.kvStateMachines[1] = s.mockStateTransition().ServiceEventsMatch(emptySlice[flow.ServiceEvent]()).Mock()
+
+	s.protocolKVStoreDB.On("IndexTx", s.candidate.ID(), expectedResultingStateID).Return(func(*transaction.Tx) error { return nil }).Once()
+	s.protocolKVStoreDB.On("StoreTx", expectedResultingStateID, &s.evolvingState).Return(func(*transaction.Tx) error { return exception }).Once()
+
+	_, dbUpdates, err := s.mutableState.EvolveState(s.candidate.ParentID, s.candidate.View, []*flow.Seal{})
+	require.NoError(s.T(), err) // `EvolveState` should succeed, because storing the encoded snapshot only happens when we execute dbUpdates
+
+	// Provide the blockID and execute the resulting `DeferredDBUpdate`. Thereby,
+	// the expected mock methods should be called, which is asserted by the testify framework
+	err = dbUpdates.Pending().WithBlock(s.candidate.ID())(&transaction.Tx{})
+	require.Error(s.T(), err)
```

Comment:
```suggestion
```

Reply:
done in https://github.com/onflow/flow-go/pull/5644/commits/9c38cfa2bb5b0d4b2e0b2dc4dd6e78c548ff2cdb